runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java

```diff
@@ -281,6 +278,10 @@ public void run(SourceContext<WindowedValue<ValueWithRecordId<OutputT>>> ctx) th
 
     ctx.emitWatermark(new Watermark(Long.MAX_VALUE));
 
+    finalizeSource();
+  }
+
+  private void finalizeSource() {
```

Comment:
Yes, the bug was that the finalize code was never executed for empty readers.
> Shall we call this method only from the 1st if statement L205 as even not, in the other else blocks will exit only after isRunning is false.

All `if` blocks will eventually execute `finalizeSource`, so I wouldn't call it directly in line 205 because that would additionally require a `return;` statement.

Reply:
Makes sense.

Reply:
>Also wait is only used as sleep here. Shall we change it to sleep to make it more evident?

Yes, good idea. The object we're waiting for, can't be notified, so sleep is better here.